h/api/storage.py

```diff
@@ -114,42 +118,52 @@ def create_annotation(request, data):
     # annotation.updated get created.
     request.db.flush()
 
-    documents = models.Document.find_or_create_by_uris(
-        request.db,
-        annotation.target_uri,
-        [u['uri'] for u in document_uri_dicts],
-        created=annotation.created,
-        updated=annotation.updated)
-
-    if documents.count() > 1:
-        document = models.merge_documents(request.db,
-                                          documents,
-                                          updated=annotation.updated)
-    else:
-        document = documents.first()
-
-    document.updated = annotation.updated
-
-    for document_uri_dict in document_uri_dicts:
-        models.create_or_update_document_uri(
-            session=request.db,
-            document=document,
-            created=annotation.created,
-            updated=annotation.updated,
-            **document_uri_dict)
-
-    for document_meta_dict in document_meta_dicts:
-        models.create_or_update_document_meta(
-            session=request.db,
-            document=document,
-            created=annotation.created,
-            updated=annotation.updated,
-            **document_meta_dict)
+    models.update_document_metadata(
+        request.db, annotation, document_meta_dicts, document_uri_dicts)
+
+    return annotation
+
+
+def update_annotation(session, id_, data):
+    """
+    Update an existing annotation and its associated document metadata.
+
+    Update the annotation identified by id_ with the given
+    data. Create, delete and update document metadata as appropriate.
+
+    :param session: the database session
+    :type session: sqlalchemy.orm.session.Session
+
+    :param id_: the ID of the annotation to be updated, this is assumed to be a
+        validated ID of an annotation that does already exist in the database
+    :type id_: string
+
+    :param data: the validated data with which to update the annotation
+    :type data: dict
+
+    :returns: the updated annotation
+    :rtype: h.api.models.Annotation
+
+    """
+    # Remove any 'document' field first so that we don't try to save it on the
+    # annotation object.
+    document = data.pop('document', None)
+
+    annotation = session.query(models.Annotation).get(id_)
+
+    for key, value in data.items():
+        setattr(annotation, key, value)
```

Comment:
One thing to be aware of here is that we need to be *really* sure that the client doesn't have the ability to manipulate the keys in the `data` dictionary beyond what we assume to be possible.

In this PR, nothing calls this function outside of tests, but let's just make sure we're aware of the issues with this kind of model assignment code, and not fall into [old traps](http://guides.rubyonrails.org/v3.2.9/security.html#mass-assignment).